source/common/memory/stats.cc

```diff
@@ -46,6 +47,47 @@ void Stats::dumpStatsToLog() {
   ENVOY_LOG_MISC(debug, "TCMalloc stats:\n{}", tcmalloc::MallocExtension::GetStats());
 }
 
+Allocator::~Allocator() {
+  {
+    absl::MutexLock guard(&mutex_);
+    terminating_ = true;
+  }
+  if (tcmalloc_thread_) {
+    tcmalloc_thread_->join();
+  }
+}
+
+void Allocator::configureBackgroundMemoryRelease() {
+  RELEASE_ASSERT(!tcmalloc_thread_, "Invalid state, tcmalloc thread has not beein initialised");
+  if (background_release_rate_ > 0) {
+    tcmalloc::MallocExtension::SetBackgroundReleaseRate(
+        tcmalloc::MallocExtension::BytesPerSecond{background_release_rate_});
+    ENVOY_LOG_MISC(info, "Configured tcmalloc with background release rate: {} bytes per second",
+                   background_release_rate_);
+    // `ProcessBackgroundActions` routine needs to be invoked for background memory release to be
+    // operative. https://github.com/google/tcmalloc/blob/master/tcmalloc/malloc_extension.h#L635
+    tcmalloc_thread_ = thread_factory_.createThread(
+        [this]() -> void { tcmallocProcessBackgroundActionsThreadRoutine(); },
+        Thread::Options{"TcmallocProcessBackgroundActions"});
+  }
+}
+
+void Allocator::tcmallocProcessBackgroundActionsThreadRoutine() {
+  ENVOY_LOG_MISC(debug, "Started tcmallocProcessBackgroundActionsThreadRoutine");
+  while (true) {
```

Comment:
This seems to me like a tight loop that AFAICT has nothing limiting it from just burning cpu e.g. it's not running once / second on a timer, etc.

As well as below

Reply:
ok, now thread will block waiting for condVar